aptos-move/aptos-vm-types/src/output.rs

```diff
@@ -0,0 +1,151 @@
+            self.events,
+            self.gas_used,
+            self.status,
+        )
+    }
+
+    pub fn write_set(&self) -> &WriteSet {
+        &self.write_set
+    }
+
+    pub fn delta_change_set(&self) -> &DeltaChangeSet {
+        &self.delta_change_set
+    }
+
+    pub fn events(&self) -> &[ContractEvent] {
+        &self.events
+    }
+
+    pub fn gas_used(&self) -> u64 {
+        self.gas_used
+    }
+
+    pub fn status(&self) -> &TransactionStatus {
+        &self.status
+    }
+
+    /// Converts VMOutput into TransactionOutput which can be used by storage
+    /// backends. During this conversion delta materialization can fail, in
+    /// which case an error is returned.
+    pub fn into_transaction_output(
+        self,
+        state_view: &impl StateView,
+    ) -> anyhow::Result<TransactionOutput, VMStatus> {
+        let (write_set, delta_change_set, events, gas_used, status) = self.unpack();
+
+        // First, check if output of transaction should be discarded or delta
+        // change set is empty. In both cases, we do not need to apply any
+        // deltas and can return immediately.
+        if status.is_discarded() || delta_change_set.is_empty() {
+            return Ok(TransactionOutput::new(write_set, events, gas_used, status));
+        }
+
+        // Try to materialize deltas and add them to the write set.
+        let mut write_set_mut = write_set.into_mut();
+        let delta_writes = delta_change_set.take_materialized(state_view)?;
+        delta_writes
+            .into_iter()
+            .for_each(|item| write_set_mut.insert(item));
+
+        let write_set = write_set_mut.freeze().map_err(|_| {
+            VMStatus::Error(
+                StatusCode::DATA_FORMAT_ERROR,
+                Some(
+                    "Failed to freeze write set when converting VMOutput to TransactionOutput"
+                        .to_string(),
+                ),
+            )
+        })?;
+
+        Ok(TransactionOutput::new(write_set, events, gas_used, status))
+    }
+
+    /// Converts VM output into transaction output which storage or state sync
+    /// understand. Extends writes with values from materialized deltas.
+    pub fn output_with_delta_writes(
+        self,
+        delta_writes: Vec<(StateKey, WriteOp)>,
+    ) -> TransactionOutput {
+        // We should have a materialized delta for every delta in the output.
+        let (write_set, delta_change_set, events, gas_used, status) = self.unpack();
+        assert_eq!(delta_writes.len(), delta_change_set.len());
+
+        let mut write_set_mut = write_set.into_mut();
+        // Add the delta writes to the write set of the transaction.
+        delta_writes
+            .into_iter()
+            .for_each(|item| write_set_mut.insert(item));
+
+        let write_set = write_set_mut
+            .freeze()
+            .expect("Freezing of WriteSet should succeed.");
+        TransactionOutput::new(write_set, events, gas_used, status)
+    }
+}
```

Comment:
testing output_with_delta_writes and into_transaction_output against each other vs ground truth would be amazing.